### PR TITLE
Support on-demand credentials in the Java binding

### PR DESCRIPTION
**Pull request by jyemin**
MONGOCRYPT-393

### CODE REVIEW

File: bindings/java/mongocrypt/src/main/java/com/mongodb/crypt/capi/MongoCryptImpl.java [hunk 1]
```diff
@@ -152,7 +158,19 @@
         }
 
         if (options.getKmsProviderOptions() != null) {
-            try (BinaryHolder binaryHolder = toBinary(options.getKmsProviderOptions())) {
+            BsonDocument kmsProviderOptions = options.getKmsProviderOptions();
+            if (options.getKmsProviderOptions().values().stream().anyMatch(value -> value.asDocument().isEmpty())) {
+
+                mongocrypt_setopt_use_need_kms_credentials_state(wrapped);
+
+                kmsProviderOptions = options.getKmsProviderOptions().entrySet().stream()
```
[the remark this answers is not in the document]
Comment (kevinAlbs) [reply bindings/java/mongocrypt/src/main/java/com/mongodb/crypt/capi/MongoCryptImpl.java:166]: Agreed. #257 has been merged.

File: bindings/java/mongocrypt/src/main/java/com/mongodb/crypt/capi/MongoCryptContextImpl.java [hunk 2]
```diff
@@ -32,8 +32,10 @@
 import static com.mongodb.crypt.capi.CAPI.mongocrypt_ctx_mongo_feed;
 import static com.mongodb.crypt.capi.CAPI.mongocrypt_ctx_mongo_op;
 import static com.mongodb.crypt.capi.CAPI.mongocrypt_ctx_next_kms_ctx;
+import static com.mongodb.crypt.capi.CAPI.mongocrypt_ctx_provide_kms_providers;
 import static com.mongodb.crypt.capi.CAPI.mongocrypt_ctx_state;
 import static com.mongodb.crypt.capi.CAPI.mongocrypt_ctx_status;
+import static com.mongodb.crypt.capi.CAPI.mongocrypt_setopt_kms_providers;
```
Comment (kevinAlbs) [inline bindings/java/mongocrypt/src/main/java/com/mongodb/crypt/capi/MongoCryptContextImpl.java:38]: Remove unused import?

Comment (kevinAlbs) [inline bindings/java/mongocrypt/src/main/java/com/mongodb/crypt/capi/MongoCryptContextImpl.java:38]: Remove unused import?

**Review by kevinAlbs — APPROVED** [review]
Changes LGTM.

https://github.com/mongodb/libmongocrypt/pull/257 has been merged. Feel free to re-request review if there are significant changes after.

File: bindings/java/mongocrypt/src/main/java/com/mongodb/crypt/capi/MongoCryptImpl.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (jyemin) [reply bindings/java/mongocrypt/src/main/java/com/mongodb/crypt/capi/MongoCryptImpl.java]: Done.

**Comment by jyemin** [comment]
Rebased on https://github.com/mongodb/libmongocrypt/pull/257

**Comment by jyemin** [comment]
I realized that this change is backwards breaking: if an older Java driver tries to use a mongodb-crypt release containing this patch, it should still work.  So I added a configuration option to MongoCrypt to make use of the new state opt-in.

**Pull request merged**